### PR TITLE
add scoring onnx

### LINKED ISSUE #1549: [FEATURE] OnnxScoringModel similar to OnnxEmbeddingModel
The reordering model ScoringModel currently only supports two types of online models. Can it be similar to OnnxEmbeddingModel to implement the use of OnnxScoringModel and local reordering models

### PR DESCRIPTION
**Pull request by hzkitty**
## Issue
Closes #1549

## Change
Added `OnnxScoringModel` similar to `OnnxEmbeddingModel`.
Documentation: https://docs.langchain4j.dev/integrations/scoring-reranking-models/in-process

## General checklist
- [ ] There are no breaking changes
- [ ] I have added unit and integration tests for my change
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

## Checklist for adding new model integration
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration
- [ ] I have added a `{NameOfIntegration}OnnxScoringModelIT` that extends from either `OnnxScoringModelIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration
- [x] I have manually verified that the `{NameOfIntegration}OnnxScoringModel` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-scoring/pom.xml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.microsoft.onnxruntime</groupId>
+            <artifactId>onnxruntime</artifactId>
+            <version>1.17.1</version>
```
Comment (langchain4j) [inline langchain4j-scoring/pom.xml]:
```suggestion
            <version>1.19.2</version>
```

File: langchain4j-scoring/pom.xml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        <artifactId>langchain4j-parent</artifactId>
+        <version>0.35.0-SNAPSHOT</version>
+        <relativePath>../langchain4j-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-scoring</artifactId>
+    <name>langchain4j-scoring</name>
+
+    <properties>
+        <ai.djl.version>0.28.0</ai.djl.version>
```
Comment (langchain4j) [inline langchain4j-scoring/pom.xml]:
```suggestion
        <ai.djl.version>0.30.0</ai.djl.version>
```

File: langchain4j-scoring/pom.xml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            <version>1.17.1</version>
+        </dependency>
+
+        <dependency>
+            <groupId>ai.djl</groupId>
+            <artifactId>api</artifactId>
+            <version>${ai.djl.version}</version>
+            <exclusions>
+                <exclusion>
+                    <!-- due to CVE-2024-26308 and CVE-2024-25710-->
```
Comment (langchain4j) [inline langchain4j-scoring/pom.xml]: worth checking if vulnerability is still in place in the latest (30) version

File: langchain4j-scoring/pom.xml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        </dependency>
+
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>9</source>
```
Comment (langchain4j) [inline langchain4j-scoring/pom.xml]: Is there a reason for version 9 specifically?

File: langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/OnnxScoringBertBiEncoder.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import ai.onnxruntime.OrtException;
+import ai.onnxruntime.OrtSession;
+import ai.onnxruntime.OrtSession.Result;
+
+import java.nio.file.Paths;
+import java.util.*;
+
+import static ai.onnxruntime.OnnxTensor.createTensor;
+
+public class OnnxScoringBertBiEncoder {
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/OnnxScoringBertBiEncoder.java]:
```suggestion
public class OnnxScoringBertCrossEncoder {
```

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+class OnnxScoringModelIT {
+
+    @Test
+    void should_score_single_text() {
+
+//        String pathToModel = "D:\\github\\langchat\\model\\reranker\\bge-reranker-large\\onnx\\model.onnx";
+//        String pathToTokenizer = "D:\\github\\langchat\\model\\reranker\\bge-reranker-large\\tokenizer.json";
+//        ScoringModel model = new OnnxScoringModel(pathToModel, new OrtSession.SessionOptions(), pathToTokenizer, 510,  true);
+
+        String pathToModel = "D:\\github\\langchat\\model\\reranker\\bge-reranker-v2-m3\\onnx\\model.onnx";
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: This won't work on CI. Can you please implement model downloading, similar to how it is done [here](https://github.com/langchain4j/langchain4j-embeddings/blob/main/langchain4j-embeddings/src/test/java/dev/langchain4j/model/embedding/onnx/OnnxEmbeddingModelIT.java)?

Comment (langchain4j) [reply langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: Perharps there is smaller model?

File: pom.xml [hunk 7]
```diff
@@ -95,6 +95,7 @@
 
         <!-- experimental -->
         <module>experimental/langchain4j-experimental-sql</module>
+        <module>langchain4j-scoring</module>
```
Comment (langchain4j) [inline pom.xml]: please add it to `langchain4j-bom/pom.xml` as well

File: langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/AbstractInProcessScoringModel.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            return new OnnxScoringBertBiEncoder(pathToModel, option, pathToTokenizer, modelMaxLength, normalize);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected abstract OnnxScoringBertBiEncoder model();
+
+    public Response<List<Double>> scoreAll(List<TextSegment> segments, String query) {
+        OnnxScoringBertBiEncoder.ScoringAndTokenCount embeddingAndTokenCount = this.model().scoreAll(query,
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/AbstractInProcessScoringModel.java]:
```suggestion
        OnnxScoringBertBiEncoder.ScoringAndTokenCount scoresAndTokenCount = this.model().scoreAll(query,
```

File: langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModel.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package dev.langchain4j.model.scoring.onnx;
+
+import ai.onnxruntime.OrtSession;
+
+public class OnnxScoringModel extends AbstractInProcessScoringModel {
+
+    private static final int DEFAULT_MODEL_MAX_LENGTH = 510; // 512 - 2 (special tokens [CLS] and [SEP])
+
+    private static final boolean DEFAULT_NORMALIZE = true;
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModel.java]:
```suggestion
    private static final boolean DEFAULT_NORMALIZE = false;
```

File: pom.xml [hunk 10]
```diff
@@ -95,6 +95,7 @@
...
 
     </modules>
```
Comment (langchain4j) [inline pom.xml:102]: would be great to add documentation page similar to `docs/docs/integrations/scoring-reranking-models/2-cohere.md`

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        ScoringModel model = new OnnxScoringModel(pathToModel, new OrtSession.SessionOptions(), pathToTokenizer, 8000,  true);
+
+        String text = "The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.";
+        String query = "What is panda?";
+
+        // when
+        Response<Double> response = model.score(text, query);
+
+        // then
+        assertThat(response.content()).isCloseTo(0.99, withPercentage(1));
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: please set `normalize=false` and add assertions with expected scores from [HF](https://huggingface.co/BAAI/bge-reranker-v2-m3)
```
scores = reranker.compute_score([['what is panda?', 'hi'], ['what is panda?', 'The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.']])
print(scores) # [-8.1875, 5.26171875]
```

**Review by ghost — COMMENTED** [review]
@1758225523 thanks a lot!

File: langchain4j-scoring/pom.xml (hunk 1 above)
Comment (hzkitty) [reply langchain4j-scoring/pom.xml]: But I see that the version of <groupId>com.microsoft.onnxruntime</groupId> in langchain4j-embeddings is still 1.17.1

File: langchain4j-scoring/pom.xml (hunk 2 above)
Comment (hzkitty) [reply langchain4j-scoring/pom.xml]: But I see that the version of <groupId>ai.djl</groupId> in langchain4j-embeddings is still 0.28.0

File: langchain4j-scoring/pom.xml (hunk 3 above)
Comment (hzkitty) [reply langchain4j-scoring/pom.xml]: But I see that the version of <groupId>ai.djl</groupId> in langchain4j-embeddings is still 0.28.0

File: langchain4j-scoring/pom.xml (hunk 4 above)
Comment (hzkitty) [reply langchain4j-scoring/pom.xml]: <groupId>org.apache.maven.plugins</groupId> I will delete it

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java (hunk 6 above)
Comment (hzkitty) [reply langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: I will use a smaller ms-marco-MiniLM-L-6-v2 model

File: pom.xml (hunk 7 above)
Comment (hzkitty) [reply pom.xml]: I will add this

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java (hunk 11 above)
Comment (hzkitty) [reply langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: I will add this

File: pom.xml (hunk 10 above)
Comment (hzkitty) [reply pom.xml:102]: I will add this

File: langchain4j-scoring/pom.xml (hunk 1 above)
Comment (langchain4j) [reply langchain4j-scoring/pom.xml:29]: The latest version is 1.19.2: https://mvnrepository.com/artifact/com.microsoft.onnxruntime/onnxruntime

Comment (langchain4j) [reply langchain4j-scoring/pom.xml]: Ok, makes sense to keep the same version, we can update it later

File: docs/docs/integrations/scoring-reranking-models/in-process.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+```java
+String pathToModel = "/home/langchain4j/model.onnx";
+String pathToTokenizer = "/home/langchain4j/tokenizer.json";
+OnnxScoringModel scoringModel = new OnnxScoringModel(pathToModel, pathToTokenizer);
+
+Response<Double> response = scoringModel.score("query", "passage");
+Double score = response.content();
+```
+
+If you want to use the GPU. onnxruntime_gpu version can be found  [here](https://huggingface.co/Xenova).
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/in-process.md]: seems to be a wrong URL

Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/in-process.md]:
```suggestion
If you want to use the GPU, onnxruntime_gpu version can be found  [here](https://huggingface.co/Xenova).
```

File: docs/docs/integrations/scoring-reranking-models/in-process.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+Many models already converted to ONNX format are available [here](https://huggingface.co/Xenova).
+
+### Usage
+
+By default, Scoring (Reranking) use the CPU. 
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-scoring</artifactId>
```
Comment (langchain4j) [inline docs/docs/integrations/scoring-reranking-models/in-process.md]:
```suggestion
    <artifactId>langchain4j-onnx-scoring</artifactId>
```
I am not sure about the name, WDYT?

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    private static ScoringModel model;
+
+    @BeforeAll
+    static void initModel() throws IOException {
+
+        // I set up a local proxy to download the model
+        // System.setProperty("https.proxyHost","127.0.0.1" );
+        // System.setProperty("https.proxyPort","7890" );
+
+        URL modelUrl = new URL("https://huggingface.co/Xenova/ms-marco-MiniLM-L-6-v2/resolve/main/onnx/model.onnx?download=true");
```
Comment (langchain4j) [inline langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: nit: what about using [model_quantized.onnx](https://huggingface.co/Xenova/ms-marco-MiniLM-L-6-v2/blob/main/onnx/model_quantized.onnx) instead? It is even smaller == faster build

**Review by ghost — COMMENTED** [review]
@1758225523 thank you!

File: docs/docs/integrations/scoring-reranking-models/in-process.md (hunk 12 above)
Comment (hzkitty) [reply docs/docs/integrations/scoring-reranking-models/in-process.md]: I will correct it

File: langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java (hunk 14 above)
Comment (hzkitty) [reply langchain4j-scoring/src/main/test/java/dev/langchain4j/model/scoring/onnx/OnnxScoringModelIT.java]: ok

**Review by ghost — APPROVED** [review]
@1758225523 thanks a lot!

**Pull request merged**